src/blocks/tipset.rs

```diff
@@ -180,6 +180,36 @@ impl Tipset {
             .transpose()?)
     }
 
+    /// Constructs and returns a full tipset if messages from storage exists
+    pub fn fill_from_blockstore(&self, store: impl Blockstore) -> Option<FullTipset> {
+        // Collect all messages before moving tipset.
+        let messages = self
+            .blocks()
+            .iter()
+            .map(|h| crate::chain::store::block_messages(&store, h))
+            .collect::<Result<Vec<_>, _>>()
+            .ok()?;
+
+        // Zip messages with blocks
+        let blocks = self
+            .blocks()
+            .iter()
+            .cloned()
+            .zip(messages)
+            .map(|(header, (bls_messages, secp_messages))| Block {
+                header,
+                bls_messages,
+                secp_messages,
+            })
+            .collect();
+
+        // the given tipset has already been verified, so this cannot fail
+        Some(
+            FullTipset::new(blocks)
+                .expect("block headers have already been verified so this check cannot fail"),
```

Comment:
nit: what about returning a `Result` here instead? Then you don't have to make an assumption you make in `expect`, because this is basically a glorified `unwrap`. 

The way this has been done in some other projects I've seen is: if you want to mark a certain code path as `defensive` - just use some sort of macro for it, that would error out in dev mode and log an error in production. Or something similar. 

I think having `expect` in production code is not the best approach to be honest.

Reply:
Logically it cannot fail. Returning a `Result` from a non-fallible function is worse than a (well-described) `expect`, imo.

Reply:
Typically `expect/panic` usages are either somewhere top-level or within tests. It's not obvious from the code above that it cannot fail either, if you want to make it obvious - do this:
```
if Ok(..) {
  return Some()
}
return None
```

That's a much better solution, than an expect to save a couple of lines of code.

Reply:
The intent is not to save a few lines of code. This function is infallible. I can prove this to another human, but I cannot (easily) prove it to rustc.

Swallowing the panic would be even worse. If this condition does not hold up then there's a bug in Forest, and we should halt.

Reply:
I think it's correct to panic here

Reply:
> This function is infallible.

Let's then make it truly infallible by skipping the block validation? Unless there is a good reason not to.

Reply:
> > This function is infallible.
> 
> Let's then make it truly infallible by skipping the block validation? Unless there is a good reason not to.

That is a possibility. But then we would lose out on an assertion. Let's keep it for now and find a good way of proving the correctness of the function at a later time.

Reply:
let's do. lgtm